cupy/core/_fusion_kernel.pyx

```diff
@@ -0,0 +1,350 @@
+            block_stride = max(
+                contiguous_size, block_size // reduce_block_size)
+            block_stride = internal.clp2(block_stride // 2 + 1)  # floor
+            block_strides.append(block_stride)
+
+        shared_mem = block_size * 32  # max bytesize of reduce_ctype.
+        return block_strides, block_size, shared_mem
+
+    cdef tuple _reduce_dims(self, list ndarray_list):
+        """Reduce number of dimensions of ndarrays and returns the cache key.
+        """
+        cdef list params = self._params
+        cdef list ndims = []
+        cdef ndarray array
+        cdef int i
+
+        for i in range(len(params)):
+            param = params[i]
+            if param.ndim <= 1:
+                continue
+            array = ndarray_list[i]
+            array = array.reduced_view()
+            ndarray_list[i] = array
+            ndims.append(array.ndim)
+
+        return tuple(ndims)
+
+    cdef list _get_inout_args(self, tuple args, list ndarray_list):
+        """Get the arguments passed to ``kern.linear_launch``.
+        """
+        cdef list params = []
+        cdef list indexers = []
+        cdef list block_strides = []
+
+        for i in range(len(self._params)):
+            array = ndarray_list[i]
+            if isinstance(array, ndarray):
+                params.append(array)
+                indexers.append(Indexer(array.shape))
+            elif self._input_index[i] >= 0:
+                scalar = args[<Py_ssize_t>self._input_index[i]]
+                params.append(scalar_to_numpy_scalar(scalar))
+
+        return params + indexers
+
+    cdef str _get_cuda_params(self, tuple key, list ndarray_list):
+        """Get a string of parameters of CUDA main function code.
+        """
+        cdef int i
+
+        if key in self._cuda_params_memo:
+            return self._cuda_params_memo[key]
+
+        cuda_params = []
+        indexers = []
+
+        for i in range(len(self._params)):
+            a = self._params[i]
+            if isinstance(a, _TraceArray):
+                ndim = ndarray_list[i].ndim
+                cuda_params.append(a.format(
+                    'CArray<${type}, ${ndim}> ${var}', ndim=ndim))
+                indexers.append(
+                    a.format('CIndexer<${ndim}> ${indexer}', ndim=ndim))
+            elif isinstance(a, _TraceScalar):
+                if a.const_value is None:
+                    cuda_params.append(a.format('${type} ${var}'))
+            else:
+                raise TypeError('Unknown type {}.'.format(type(a)))
+
+        ret = cuda_params + indexers + self._block_strides
+        ret = ', '.join(ret)
+        self._cuda_params_memo[key] = ret
+        return ret
+
+    def execute(self, tuple args, list shapes):
+        ndarray_list = self._get_ndarray_list(args, shapes)
+        ret = self._get_return_value(ndarray_list)
+        size_info = self._get_kernel_size(ndarray_list)
```

Comment:
`size_info` is too ambiguous, too. (I think this temporary variable is not needed)